File: tests/07_phases_ompss.dg/c/success_taskloop_04.c
/*--------------------------------------------------------------------
  (C) Copyright 2006-2012 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/



/*
<testinfo>
test_generator=config/mercurium-ompss
</testinfo>
*/
#include <stdlib.h>
#include <stdio.h>

enum { N = 10000 };

char check[N] = { };

void f(int *a, int n)
{
    int i;
    #pragma omp taskloop out(a[i]) grainsize(10) shared(check) nogroup
    for (i = 0; i < n; i++)
    {
        if (check[i] != 0)
            abort();
        a[i] = i;
        check[i] = 1;
    }

#pragma omp taskloop inout(a[i]) grainsize(10) shared(check) nogroup
    for (i = 0; i < n; i++)
    {
        if (check[i] != 1)
            abort();
        a[i]++;
        check[i] = 0;
    }

#pragma omp taskloop in(a[i]) grainsize(10) shared(check)
    for (i = 0; i < n; i++)
    {
        if (check[i] != 0)
            abort();
    }
}

int w[N];

int main(int argc, char* argv[])
{
    int i;
    for (i = 0; i < N; i++)
    {
        w[i] = i;
    }

    f(w, N);

    return 0;
}

File: tests/07_phases_ompss.dg/fortran/success_taskloop_04.f90
! <testinfo>
! test_generator=config/mercurium-ompss
! </testinfo>
PROGRAM P
    IMPLICIT NONE
    INTEGER :: G, I
    INTEGER, PARAMETER :: MAX_GRAINSIZE = 20
    INTEGER, PARAMETER :: N = 2000
    INTEGER :: A(N)

    INTEGER :: L, U, S

    
    A = 0
    L = 1
    U = 2000
    S = 1
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -1


    A = 0
    L = 1
    U = 2000
    S = 3
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -2


    A = 0
    L = 2000
    U = 1
    S = -1
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -3


    A = 0
    L = 2000
    U = 1
    S = -3
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -4


    A = 0
    L = -999
    U = 1000
    S = 1
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(I+1000) = A(I+1000) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -5


    A = 0
    L = -999
    U = 1000
    S = 3
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(I+1000) = A(I+1000) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -6


    A = 0
    L = 999
    U = -1000
    S = -1
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(1000 - I) = A(1000 - I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -7


    A = 0
    L = 999
    U = -1000
    S = -3
    DO G=1, MAX_GRAINSIZE

        !$OMP TASKLOOP SHARED(A) GRAINSIZE(G)
        DO I=L, U, S
            A(1000 - I) = A(1000 - I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_GRAINSIZE)) STOP -8

END PROGRAM P
